### ISSUE TITLE
Idea - NNSVS Support

### ISSUE THREAD

**Issue by SeleDreams** [titinko/utsu#69]
It's an idea of something i'm thinking about working on but am not sure about the interest.
I have the idea to integrate a secondary engine to utsu named NNSVS which is an engine for AI based vocal synthesis that is open source https://github.com/r9y9/nnsvs it doesn't support the same type of voicebanks as they require to be AI trained but i feel like having one editor to use multiple types of voicebanks would be great, even more for voicebanks of better quality AI trained

NNSVS currently doesn't have compatible engine so i felt like it would also help and bring more people to utsu and nnsvs as well, helping the open source community

**Comment by SeleDreams** [comment]
It would allow UTSU to become kind of the "AI UTAU" for the community where anyone could create AI vocals and make music with them

**Comment by Luk3M** [comment]
I think this is a great idea! Would it generate a voicebank for UTSU's resample, or it also works as a whole resampler?

**Comment by adlez27** [comment]
This might be helpful: https://note.com/crazy_utau/n/n45db22b33d2c

**Comment by SeleDreams** [comment]
> This might be helpful: https://note.com/crazy_utau/n/n45db22b33d2c

ENUNU is separate because it relies on the UTAU plugin api from what i remember it doesn't use the same system at all than utau in reality and relies on the data it can access to via plugins, that's why in the git it specifies the things it couldn't do due to the limitations of the plugin api 



> I think this is a great idea! Would it generate a voicebank for UTSU's resample, or it also works as a whole resampler?

it works as its own thing due to the difference of AI synthesis

**Comment by Luk3M** [comment]
Reading about the software, it looks like the task would be then to make UTSU be able to export to MusicXML so NNSVS can interpret it, right? And maybe bundle it in if the license permits.
This could be the open source version of [Synthesizer V AI](https://synthv.fandom.com/wiki/Synthesizer_V_AI), which makes me very excited about!

**Comment by SeleDreams** [comment]
> Reading about the software, it looks like the task would be then to make UTSU be able to export to MusicXML so NNSVS can interpret it, right? And maybe bundle it in if the license permits.
> This could be the open source version of [Synthesizer V AI](https://synthv.fandom.com/wiki/Synthesizer_V_AI), which makes me very excited about!

it technically could but I think some small editing would be enough for it to directly get the data from UST files, all it needs is the data, the files it gets the data from aren't as important as long as there are deserializers

and yes this is one of the goals

**Comment by SeleDreams** [comment]
licence wise, nnsvs is MIT, there is 0 problems with bundling it as from what i can see UTSU is under MIT as well

**Comment by titinko** [comment]
Times like this remind me how much UTSU needs its own plugin framework.

If I understand correctly, integrating NNSVS with UTSU would have three parts to it.

Rendering songs in NNSVS: The easiest way is to write code converting the internal Song object into an NNSVS file, then run NNSVS on that in the background.

Using NNSVS voicebanks: I could see UTSU's song editor being tweaked so that it pretends that NNSVS voicebanks are regular UTAU voicebanks on the frontend, but in the backend only renders them with NNSVS.

Creating NNSVS voicebanks: Since the format is completely different from UTAU's voicebanks, you'd have to write an entirely new voicebank editor UI.

**Comment by SeleDreams** [comment]
> Times like this remind me how much UTSU needs its own plugin framework.
> 
> If I understand correctly, integrating NNSVS with UTSU would have three parts to it.
> 
> **Rendering songs in NNSVS:** The easiest way is to write code converting the internal Song object into an NNSVS-readable file, then run NNSVS on that in the background.
> 
> **Using NNSVS voicebanks:** I could see UTSU's song editor being tweaked so that it pretends that NNSVS voicebanks are regular UTAU voicebanks on the frontend, but in the backend only renders them with NNSVS.
> 
> **Creating NNSVS voicebanks:** Since the format is completely different from UTAU's voicebanks, you'd have to write an entirely new voicebank editor UI.

Yes, I imagine that the vb creation side would come last since it's not the priority, usage would come first and vb creation second (since creating AI vbs to begin with is much more complex due to the AI training phase)

**Comment by Luk3M** [comment]
I wanna try making a serializer from the Song object to NNSVS. Looks like there's a need to convert it to HTS full-context label files (using Sinsy), and NNSVS makes a MusicXML to label file step to use it.

If i'm correct, does anyone knows any document with the specification of the data from HTS label files?
